### PR TITLE
Improve reduce sum

### PR DESCRIPTION
**Pull request by emricksinisonos**

### CODE REVIEW

File: core/src/ops/nn/reduce.rs [hunk 1]
```diff
@@ -142,44 +143,58 @@ impl Reducer {
...
+        f32: AsPrimitive<T>,
     {
         if axes.len() == 0 {
             return input.to_owned();
         }
+
         let mut output: Option<ArrayD<T>> = None;
-        for axis in axes.iter() {
-            let current_input = output
+        for axis in axes.iter().cloned() {
```
Comment (kali) [inline core/src/ops/nn/reduce.rs:155]: Nitpicking: can we dispense with the cloned() ? These are just usize, right ? so `&axis` or at least `.copied()` should be enough?

File: linalg/src/frame/reduce/mod.rs [hunk 2]
```diff
@@ -280,7 +249,7 @@ pub mod test {
         let found = values;
         let red = op.run_with_params(found, params).unwrap();
         tensor0(red)
-            .close_enough(&tensor0(expected), true)
+            .close_enough(&tensor0(expected), Approximation::SuperApproximate)
```
Comment (kali) [inline linalg/src/frame/reduce/mod.rs:252]: do we still need SuperApprox when we use integer-as-float inputs ?

File: linalg/src/frame/reduce/sum.rs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    use proptest::test_runner::TestCaseResult;
+
+    #[macro_export]
+    macro_rules! sum_frame_tests {
+        ($cond:expr, $t: ty, $ker:ty) => {
+            proptest::proptest! {
+                #[test]
+                // Addition is not associative in f16 a + b + c != (a + b) + c because of precision
+                // issues. To avoid that, test vector should be aligned with kernel size (to avoid
+                // having to sum partial sums together).
```
Comment (kali) [inline linalg/src/frame/reduce/sum.rs:16]: is this still true ? it is a bit misleading in that f32 is not associative either.

File: linalg/src/generic/reduce.rs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    #[cfg(test)]
+    #[macro_use]
+    pub mod h {
+        use super::*;
+        crate::sum_frame_tests!(true, f16, crate::generic::reduce::sum::HSum8);
+    }
+}
+
+// Softmax generic implementation
+pub mod softmax {
```
Comment (kali) [inline linalg/src/generic/reduce.rs:109]: Should this be softmatxl2 instead ?

**Review by kali — APPROVED** [review]
Looks good, just a couple of nitpicking.

File: core/src/ops/nn/reduce.rs (hunk 1 above)
Comment (emricksinisonos) [reply core/src/ops/nn/reduce.rs:155]: Yes, I just didn't want to dereference axis everywhere in the scope. I'll change that to `.copied()`.

File: linalg/src/frame/reduce/mod.rs (hunk 2 above)
Comment (emricksinisonos) [reply linalg/src/frame/reduce/mod.rs:252]: You're right, it should be fine with `Approximate`

**Pull request merged**